File: skills/doit-mcp-api/examples/query-costs-by-service.ts
/**
 * DoiT API Example: Query Costs by Service
 *
 * This example demonstrates how to query the DoiT Platform Analytics API
 * to get costs grouped by service for the last 30 days.
 *
 * Usage:
 *   DOIT_API_KEY=your_api_key npx ts-node query-costs-by-service.ts
 */

interface DoitQueryConfig {
  config: {
    dataSource: string;
    metric: { type: string; value: string };
    timeRange: { mode: string; amount: number; unit: string; includeCurrent: boolean };
    group: Array<{
      id: string;
      type: string;
      limit: { metric: { type: string; value: string }; sort: string; value: number };
    }>;
  };
}

interface QueryResults {
  schema: Array<{ name: string; type: string }>;
  rows: any[][];
}

async function getCostsByService(apiKey: string): Promise<QueryResults> {
  const queryConfig: DoitQueryConfig = {
    config: {
      dataSource: 'billing',
      metric: { type: 'basic', value: 'cost' },
      timeRange: {
        mode: 'last',
        amount: 30,
        unit: 'day',
        includeCurrent: false
      },
      timeInterval: 'month',
      group: [{
        id: 'service_description',
        type: 'fixed',
        limit: {
          metric: { type: 'basic', value: 'cost' },
          sort: 'desc',
          value: 25
        }
      }]
    }
  };

  const response = await fetch('https://api.doit.com/analytics/v1/reports/query', {
    method: 'POST',
    headers: {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(queryConfig)
  });

  if (!response.ok) {
    throw new Error(`DoiT API error: ${response.status} ${response.statusText}`);
  }

  return response.json();
}

// Main execution
async function main() {
  const apiKey = process.env.DOIT_API_KEY;
  if (!apiKey) {
    console.error('Error: DOIT_API_KEY environment variable is required');
    process.exit(1);
  }

  try {
    const results = await getCostsByService(apiKey);

    console.log('Costs by Service (Last 30 Days):');
    console.log('================================');

    const costIdx = results.schema.findIndex(col => col.name === 'cost');
    results.rows.forEach(row => {
      const service = row[0];
      const cost = row[costIdx] as number;
      console.log(`${service}: $${cost.toFixed(2)}`);
    });
  } catch (error) {
    console.error('Failed to query DoiT API:', error);
    process.exit(1);
  }
}

main();

File: skills/doit-mcp-api/examples/query-costs-by-service.py
#!/usr/bin/env python3
"""
DoiT API Example: Query Costs by Service

This example demonstrates how to query the DoiT Platform Analytics API
to get costs grouped by service for the last 30 days.

Usage:
    DOIT_API_KEY=your_api_key python query-costs-by-service.py
"""

import os
import sys
import requests
from typing import Any


def get_costs_by_service(api_key: str) -> dict[str, Any]:
    """Query DoiT API for costs grouped by service."""

    query_config = {
        'config': {
            'dataSource': 'billing',
            'metric': {'type': 'basic', 'value': 'cost'},
            'timeRange': {
                'mode': 'last',
                'amount': 30,
                'unit': 'day',
                'includeCurrent': False
            },
            'timeInterval': 'month',
            'group': [{
                'id': 'service_description',
                'type': 'fixed',
                'limit': {
                    'metric': {'type': 'basic', 'value': 'cost'},
                    'sort': 'desc',
                    'value': 25
                }
            }]
        }
    }

    response = requests.post(
        'https://api.doit.com/analytics/v1/reports/query',
        headers={
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/json'
        },
        json=query_config
    )
    response.raise_for_status()
    return response.json()


def main():
    api_key = os.environ.get('DOIT_API_KEY')
    if not api_key:
        print('Error: DOIT_API_KEY environment variable is required', file=sys.stderr)
        sys.exit(1)

    try:
        results = get_costs_by_service(api_key)

        print('Costs by Service (Last 30 Days):')
        print('=' * 40)

        schema = results.get('schema', [])
        cost_idx = next(i for i, col in enumerate(schema) if col['name'] == 'cost')
        for row in results.get('rows', []):
            service = row[0]
            cost = row[cost_idx]
            print(f'{service}: ${cost:.2f}')

    except requests.RequestException as e:
        print(f'Failed to query DoiT API: {e}', file=sys.stderr)
        sys.exit(1)


if __name__ == '__main__':
    main()
